src/engines/ioda/src/ioda/Engines/ObsStore/Attributes.cpp

```diff
@@ -55,14 +55,36 @@ std::shared_ptr<Attribute> Attribute::write(gsl::span<const char> data,
   }
   Selection m_select(start, npoints);
   Selection f_select(start, npoints);
-  attr_data_->write(data, m_select, f_select, isFill);
+
+  if (dtype == *dtype_) {
```

Comment:
This code block is repeated (with some variation) several times. Is it worth using a common utility function to reduce duplication?

Reply:
It's varied four times, and there shouldn't be any more cases. Twice for writes, twice for reads. Attributes and datasets have slightly different semantics, as attributes do not support partial writes/reads. Reads vs writes have similar operations, though they in a different order (write: copy data to temp buffer, convert, then pass along to the internal storage object; read: get from internal storage, convert, then copy to client buffer). Each case has 12 lines of functional code, and I'm not sure if there is much savings overall.